### ISSUE TITLE
Why does Statement::ExecutePrepared() silently fail?

### ISSUE THREAD

**Issue by bangusi** [vrogier/ocilib#234]
Ocilib: 4.6.4 ( ANSI build )
Visual Studio 2017

This is the dataset I want to insert into a table using the array interface:

```
ATOMIC_NUMBER	ELEMENT_NAME	IS_METAL	MELTING_POINT	NUMBER_OF_VALENCE
1	Hydrogen		14.175	1
2	Helium			
3	Lithium	yes	453.85	1

```
This is the code:

```
void insert(ocilib::Connection& db_connection)
{
	std::string ins = R"(insert into ptable(
atomic_number,
element_name,
is_metal,
melting_point,
number_of_valence)
values(:1, :2, :3, :4, :5))";

	ocilib::Statement stmt(db_connection);
	stmt.Prepare(ins);
	const int array_size = 3;
	stmt.SetBindArraySize(array_size);

	int column_count = 5;
	std::vector<int> atom_nbr;
	std::vector<std::string> elem_name;
	std::vector<std::string> is_metal;
	std::vector<double> melt_pt;
	std::vector<int> nbr_valence;

	stmt.Bind(":1", atom_nbr, ocilib::BindInfo::In);
	stmt.Bind(":2", elem_name, 20, ocilib::BindInfo::In);
	stmt.Bind(":3", is_metal, 5, ocilib::BindInfo::In);
	stmt.Bind(":4", melt_pt, ocilib::BindInfo::In);
	stmt.Bind(":5", nbr_valence, ocilib::BindInfo::In);

	{
		// add some data;
		atom_nbr.push_back(1);
		atom_nbr.push_back(2);
		atom_nbr.push_back(3);

		elem_name.push_back("Hydrogen");
		elem_name.push_back("Helium");
		elem_name.push_back("Lithium");

		stmt.GetBind(":3").SetDataNull(true, 1);
		stmt.GetBind(":3").SetDataNull(true, 2);
		is_metal.push_back("yes");

		melt_pt.push_back(14.175);
		stmt.GetBind(":4").SetDataNull(true, 2);
		melt_pt.push_back(453.85);

		nbr_valence.push_back(1);
		stmt.GetBind(":5").SetDataNull(true, 2);
		nbr_valence.push_back(1);

		stmt.ExecutePrepared();
	}
	db_connection.Commit();
}
```
The program silently crashes (and no exception thrown )and no rows are inserted into target table.

**Comment by vrogier** [comment]
Hi,

for array binds, binded vectors size must be equal or greater than the value passed to Statement::SetBindArraySize().
In your example, is_metal vector contains 1 element and melt_pt/nbr_valence contain 2 elements.
OCILIB expects that these vector contain at least 3 elements.
Calling BindInfo::SetDataNull() only indicates to Oracle client that the element at pos X is null and its value shall be interpreted as NULL.

In your code, for example:

```c++
melt_pt.push_back(14.175);
		stmt.GetBind(":4").SetDataNull(true, 2);
		melt_pt.push_back(453.85);
```

you provide a vector with 2 elements and say to Oracle element #2 is null.
Thus, value '14.175' will be used for 1st element, NULL will be used for 2nd element instead of '453.85' 

Thus, make sure you have the right number of elements in your vector. 

Regards,

Vincent

**Comment by bangusi** [comment]
Ok. Now I do (pre-sizing the vectors ) but still failing . This is a bit strange because I thought I have used this interface successfully before. Please, can you run this code?

```
create table ptable
(
atomic_number number(*,0) not null enable,
element_name  varchar2(32) not null enable,
is_metal  varchar2(5) not null enable,
melting_point number(*,0) not null enable,
number_of_valence number(*,0) not null enable
);
```
```
void insert(ocilib::Connection& db_connection)
{
std::string ins = R"(insert into ptable(
atomic_number,
element_name,
is_metal,
melting_point,
number_of_valence)
values(:1, :2, :3, :4, :5))";

	ocilib::Statement stmt(db_connection);
	stmt.Prepare(ins);
	const int array_size = 3;
	stmt.SetBindArraySize(array_size);

	int column_count = 5;
	std::vector<int> atom_nbr(array_size);
	std::vector<std::string> elem_name(array_size);
	std::vector<std::string> is_metal(array_size);
	std::vector<double> melt_pt(array_size);
	std::vector<int> nbr_valence(array_size);

	stmt.Bind(":1", atom_nbr, ocilib::BindInfo::In);
	stmt.Bind(":2", elem_name, 20, ocilib::BindInfo::In);
	stmt.Bind(":3", is_metal, 5, ocilib::BindInfo::In);
	stmt.Bind(":4", melt_pt, ocilib::BindInfo::In);
	stmt.Bind(":5", nbr_valence, ocilib::BindInfo::In);

	{
		// add some data;
		atom_nbr[0] = 1;
		atom_nbr[1] = 2;
		atom_nbr[2] = 3;

		elem_name[0] = "Hydrogen";
		elem_name[1] = "Helium";
		elem_name[2] = "Lithium";

		stmt.GetBind(":3").SetDataNull(true, 1);
		stmt.GetBind(":3").SetDataNull(true, 2);
		is_metal[2] = "yes";

		melt_pt[0] = 14.175;
		stmt.GetBind(":4").SetDataNull(true, 2);
		melt_pt[2] = 453.85;

		nbr_valence[0] = 1;
		stmt.GetBind(":5").SetDataNull(true, 2);
		nbr_valence[2] = 1;

		assert(atom_nbr.size() == array_size
			&& elem_name.size() == array_size
			&& elem_name.size() == array_size
			&& is_metal.size() == array_size
			&& melt_pt.size() == array_size
			&& nbr_valence.size() == array_size );

		stmt.ExecutePrepared();
	}
	db_connection.Commit();
}
```

**Comment by vrogier** [comment]
Hi,

I get an error ORA-24381: error(s) in array DML.
Which is logical as you have declared the  table columns as not null.

**Comment by vrogier** [comment]
Here is your code with a try catch block retrieving batch errors:
```c++
void main()
{
    try
    {
        Environment::Initialize(Environment::Default);

        con.Open("db18c", "usr", "pwd", Environment::SessionDefault);
        std::string sql = R"(insert into ptable(
                            atomic_number,
                            element_name,
                            is_metal,
                            melting_point,
                            number_of_valence)
                            values(:1, :2, :3, :4, :5))";

        Statement stmt(con);
        stmt.Prepare(sql);
        const int array_size = 3;
        stmt.SetBindArraySize(array_size);

        int column_count = 5;
        std::vector<int> atom_nbr(array_size);
        std::vector<std::string> elem_name(array_size);
        std::vector<std::string> is_metal(array_size);
        std::vector<double> melt_pt(array_size);
        std::vector<int> nbr_valence(array_size);

        stmt.Bind(":1", atom_nbr, ocilib::BindInfo::In);
        stmt.Bind(":2", elem_name, 20, ocilib::BindInfo::In);
        stmt.Bind(":3", is_metal, 5, ocilib::BindInfo::In);
        stmt.Bind(":4", melt_pt, ocilib::BindInfo::In);
        stmt.Bind(":5", nbr_valence, ocilib::BindInfo::In);

        atom_nbr[0] = 1;
        atom_nbr[1] = 2;
        atom_nbr[2] = 3;

        elem_name[0] = "Hydrogen";
        elem_name[1] = "Helium";
        elem_name[2] = "Lithium";

        stmt.GetBind(":3").SetDataNull(true, 1);
        stmt.GetBind(":3").SetDataNull(true, 2);
        is_metal[2] = "yes";

        melt_pt[0] = 14.175;
        stmt.GetBind(":4").SetDataNull(true, 2);
        melt_pt[2] = 453.85;

        nbr_valence[0] = 1;
        stmt.GetBind(":5").SetDataNull(true, 2);
        nbr_valence[2] = 1;

        try
        {
            stmt.ExecutePrepared();
        }
        catch (Exception& ex)
        {
            std::vector<Exception> batchExceptions;
            stmt.GetBatchErrors(batchExceptions);
            for (auto& batchEx : batchExceptions)
            {
                ocout << "at row " << batchEx.GetRow() << " error: " << batchEx.what() << oendl;
            }
        }
        con.Commit();

    }
    catch (Exception& ex)
    {
        ocout << ex.what() << oendl;
    }

    Environment::Cleanup();
}
```

gives the output :

```bash
at row 1 error: ORA-01400: cannot insert NULL into ("USR"."PTABLE"."IS_METAL")
at row 2 error: ORA-01400: cannot insert NULL into ("USR"."PTABLE"."IS_METAL")
```

**Comment by vrogier** [comment]
In the code I have posted, you will notice the usage of a nested try/catch block around the  ExecutePrepared() call. If I would have placed the batch error retrieval in the outer catch block, it would have failed to retrieve the error information as the stack being unwinded, the destructor of the stmt variable would have freed the underlying C Statement handle leading ex.GetStatement() returning a dandling C Statement handle pointer to recreate a C++ Statement object to access its GetBatchErrors() method. 
To summarize, if you want to access to Exception::GetStatement(), or Exception::GetConnection() or even to the Statement variable that generated an error, make sure the C++ object is declared outside the try/catch block. As Ocilib C++ API objects are performing RAII with reference counting, their destructor frees underlying resources prior getting in catch blocks.

Regards,

Vincent

**Comment by bangusi** [comment]
@vrogier Thanks. There is a lot I didn't know. That `ORA-24381: error(s) in array DML` has always bugged me. I want to know which row is failing.

**Comment by bangusi** [comment]
@vrogier 
While I have your attention. How do to re-adjust the array size.
Contrived example here. You have 5 rows that you want to load in batches of 3.

ATOMIC_NUMBER	ELEMENT_NAME	IS_METAL	MELTING_POINT	NUMBER_OF_VALENCE
1	Hydrogen		14.175	1
2	Helium			
3	Lithium	yes	453.85	1
4	Beryllium	yes	1560.15	2
5	Boron		2573.15	2


My ddl modified to allow nulls

```
CREATE TABLE PTABLE
(	
ATOMIC_NUMBER NUMBER(*,0) NOT NULL, 
ELEMENT_NAME VARCHAR2(32) NOT NULL, 
IS_METAL VARCHAR2(5), 
MELTING_POINT NUMBER(*,0), 
NUMBER_OF_VALENCE NUMBER(*,0)
)

```
The code:

```
After loading the first batch the second batch has 2 rows remaining.
This is how I am doing it but I am not getting the desired result.
The final batch is loading incorrectly

void insert()
{
	std::string ins = R"(insert into z_ptable(
atomic_number,
element_name,
is_metal,
melting_point,
number_of_valence)
values(:1, :2, :3, :4, :5))";

	ocilib::Statement stmt(db_connection);
	stmt.AllowRebinding(true);
	stmt.Prepare(ins);
	const int array_size = 3; // initial array size
	stmt.SetBindArraySize(array_size);

	int column_count = 5;
	std::vector<int> atom_nbr(array_size);
	std::vector<std::string> elem_name(array_size);
	std::vector<std::string> is_metal(array_size);
	std::vector<double> melt_pt(array_size);
	std::vector<int> nbr_valence(array_size);

	stmt.Bind(":1", atom_nbr, ocilib::BindInfo::In);
	stmt.Bind(":2", elem_name, 20, ocilib::BindInfo::In);
	stmt.Bind(":3", is_metal, 5, ocilib::BindInfo::In);
	stmt.Bind(":4", melt_pt, ocilib::BindInfo::In);
	stmt.Bind(":5", nbr_valence, ocilib::BindInfo::In);

	//first batch
	{
		// add some data;
		atom_nbr[0] = 1;
		atom_nbr[1] = 2;
		atom_nbr[2] = 3;

		elem_name[0] = "Hydrogen";
		elem_name[1] = "Helium";
		elem_name[2] = "Lithium";

		stmt.GetBind(":3").SetDataNull(true, 1);
		stmt.GetBind(":3").SetDataNull(true, 2);
		is_metal[2] = "yes";

		melt_pt[0] = 14.175;
		stmt.GetBind(":4").SetDataNull(true, 2);
		melt_pt[2] = 453.85;

		nbr_valence[0] = 1;
		stmt.GetBind(":5").SetDataNull(true, 2);
		nbr_valence[2] = 1;

		try
		{
			stmt.ExecutePrepared();
		}
		catch (ocilib::Exception& ex)
		{
			std::vector<ocilib::Exception> batchExceptions;
			stmt.GetBatchErrors(batchExceptions);
			for (auto& batchEx : batchExceptions)
			{
				std::cout << "at row " << batchEx.GetRow() << " error: " << batchEx.what() << std::endl;
			}
		}
	}
	//final batch
	{
		// add some data;
		atom_nbr[0] = 4;
		atom_nbr[1] = 5;

		elem_name[0] = "Beryllium";
		elem_name[1] = "Boron";

		is_metal[0] = "yes";
		stmt.GetBind(":3").SetDataNull(true, 2);

		melt_pt[0] = 1560.15;
		melt_pt[1] = 2573.15;

		nbr_valence[0] = 2;
		nbr_valence[1] = 2;

		int new_array_size = 2;  // revised array size
		stmt.SetBindArraySize(new_array_size);

		try
		{
			stmt.ExecutePrepared();
		}
		catch (ocilib::Exception& ex)
		{
			std::vector<ocilib::Exception> batchExceptions;
			stmt.GetBatchErrors(batchExceptions);
			for (auto& batchEx : batchExceptions)
			{
				std::cout << "at row " << batchEx.GetRow() << " error: " << batchEx.what() << std::endl;
			}
		}
	}
	db_connection.Commit();
}
```

**Comment by vrogier** [comment]
Hi,

Here is your code (that I modified a bit):

```c++
void main()
{
    try
    {
        Environment::Initialize(Environment::Default);

        con.Open("db18c", "usr", "pwd", Environment::SessionDefault);
        std::string sql = R"(insert into ptable values(:1, :2, :3, :4, :5))";

        auto executeStmt = [](Statement &stmt, int currentArraySize)
        {
            try
            {
                stmt.SetBindArraySize(currentArraySize);
                stmt.ExecutePrepared();
                std::cout << "row inserted : " << stmt.GetAffectedRows() << std::endl;
            }
            catch (Exception& ex)
            {
                std::vector<Exception> batchExceptions;
                stmt.GetBatchErrors(batchExceptions);
                for (auto& batchEx : batchExceptions)
                {
                    std::cout << "at row " << batchEx.GetRow() << " error: " << batchEx.what() << std::endl;
                }
            }
        };

        const int maxArraySize = 3;

        std::vector<int> atom_nbr(maxArraySize);
        std::vector<std::string> elem_name(maxArraySize);
        std::vector<std::string> is_metal(maxArraySize);
        std::vector<double> melt_pt(maxArraySize);
        std::vector<int> nbr_valence(maxArraySize);

        Statement stmt(con);
        stmt.Prepare(sql);
        stmt.SetBindArraySize(maxArraySize);

        stmt.Bind(":1", atom_nbr, ocilib::BindInfo::In);
        stmt.Bind(":2", elem_name, 20, ocilib::BindInfo::In);
        stmt.Bind(":3", is_metal, 5, ocilib::BindInfo::In);
        stmt.Bind(":4", melt_pt, ocilib::BindInfo::In);
        stmt.Bind(":5", nbr_valence, ocilib::BindInfo::In);

        // first batch

        atom_nbr[0] = 1;
        atom_nbr[1] = 2;
        atom_nbr[2] = 3;

        elem_name[0] = "Hydrogen";
        elem_name[1] = "Helium";
        elem_name[2] = "Lithium";

        stmt.GetBind(":3").SetDataNull(true, 1);
        stmt.GetBind(":3").SetDataNull(true, 2);
        is_metal[2] = "yes";

        melt_pt[0] = 14.175;
        stmt.GetBind(":4").SetDataNull(true, 2);
        melt_pt[2] = 453.85;

        nbr_valence[0] = 1;
        stmt.GetBind(":5").SetDataNull(true, 2);
        nbr_valence[2] = 1;

        executeStmt(stmt, 3);

        // second batch

        atom_nbr[0] = 4;
        atom_nbr[1] = 5;

        elem_name[0] = "Beryllium";
        elem_name[1] = "Boron";

        is_metal[0] = "yes";
        stmt.GetBind(":3").SetDataNull(true, 2);

        melt_pt[0] = 1560.15;
        melt_pt[1] = 2573.15;

        nbr_valence[0] = 2;
        nbr_valence[1] = 2;

        executeStmt(stmt, 2);

        // select inserted data

        stmt.Execute("select * from ptable");
        auto rs = stmt.GetResultset();
        while (rs++)
        {
            std::cout
                << std::left << std::setw(10) << rs.Get<std::string>(1)
                << std::left << std::setw(15) << rs.Get<std::string>(2)
                << std::left << std::setw(10) << rs.Get<std::string>(3)
                << std::left << std::setw(10) << rs.Get<std::string>(4)
                << std::left << std::setw(10) << rs.Get<std::string>(5)
                << std::endl;
        }

        con.Commit();

    }
    catch (Exception& ex)
    {
        ocout << ex.what() << oendl;
    }

    Environment::Cleanup();
}
```

which give the output:

```bash
row inserted : 3
row inserted : 2

1         Hydrogen                 14        1
2         Helium
3         Lithium        yes       454       1
4         Beryllium                1560      2
5         Boron
```

and the result in SQL*PLUS:

```sql
SQL> select * from ptable;

ATOMIC_NUMBER ELEMENT_NAME                     IS_ME MELTING_POINT NUMBER_OF_VALENCE
------------- -------------------------------- ----- ------------- -----------------
            1 Hydrogen                                          14                 1
            2 Helium
            3 Lithium                          yes             454                 1
            4 Beryllium                                       1560                 2
            5 Boron

SQL>
```

**Comment by bangusi** [comment]
The last two rows in your example are inserted incorrectly. No?
The number of rows inserted is correct but the values are not.
I am not seeing this

```
ATOMIC_NUMBER	ELEMENT_NAME	IS_METAL	MELTING_POINT	NUMBER_OF_VALENCE
4	Beryllium	yes	1560.15	2
5	Boron		2573.15	2
```

**Comment by vrogier** [comment]
arf, You're right, it is because, once you've called, BindInfo::SetDataNull(true, pos), this information is retained. You must revert it with BindInfo::SetDataNull(false, pos)

**Comment by vrogier** [comment]
modified exampe:

```c++
void main()
{
    try
    {
        Environment::Initialize(Environment::Default);

        con.Open("db18c", "usr", "pwd", Environment::SessionDefault);
        std::string sql = R"(insert into ptable values(:1, :2, :3, :4, :5))";

        auto setBindValue = [](Statement& st, int bindIndex, auto& elems, int elemIndex, auto&& value)
        {
            elems[elemIndex-1] = value;
            st.GetBind(bindIndex).SetDataNull(false, elemIndex);
        };

        auto setBindNull = [](Statement& st, int bindIndex, int elemIndex)
        {
            st.GetBind(bindIndex).SetDataNull(true, elemIndex);
        };

        auto executeStmt = [](Statement &stmt, int currentArraySize)
        {
            try
            {
                stmt.SetBindArraySize(currentArraySize);
                stmt.ExecutePrepared();
                std::cout << "row inserted : " << stmt.GetAffectedRows() << std::endl;
            }
            catch (Exception& ex)
            {
                std::vector<Exception> batchExceptions;
                stmt.GetBatchErrors(batchExceptions);
                for (auto& batchEx : batchExceptions)
                {
                    std::cout << "at row " << batchEx.GetRow() << " error: " << batchEx.what() << std::endl;
                }
            }
        };

        const int maxArraySize = 3;

        std::vector<int> atom_nbr(maxArraySize);
        std::vector<std::string> elem_name(maxArraySize);
        std::vector<std::string> is_metal(maxArraySize);
        std::vector<double> melt_pt(maxArraySize);
        std::vector<int> nbr_valence(maxArraySize);

        Statement stmt(con);
        stmt.Prepare(sql);
        stmt.SetBindArraySize(maxArraySize);

        stmt.Bind(":1", atom_nbr, ocilib::BindInfo::In);
        stmt.Bind(":2", elem_name, 20, ocilib::BindInfo::In);
        stmt.Bind(":3", is_metal, 5, ocilib::BindInfo::In);
        stmt.Bind(":4", melt_pt, ocilib::BindInfo::In);
        stmt.Bind(":5", nbr_valence, ocilib::BindInfo::In);

        // first batch

        setBindValue(stmt, 1, atom_nbr, 1, 1);
        setBindValue(stmt, 1, atom_nbr, 2, 2);
        setBindValue(stmt, 1, atom_nbr, 3, 3);

        setBindValue(stmt, 2, elem_name, 1, "Hydrogen");
        setBindValue(stmt, 2, elem_name, 2, "Helium");
        setBindValue(stmt, 2, elem_name, 3, "Lithium");

        setBindNull(stmt, 3,  1);
        setBindNull(stmt, 3, 2);
        setBindValue(stmt, 3, is_metal, 3, "yes");

        setBindValue(stmt, 4, melt_pt, 1, 14.175);
        setBindNull(stmt, 4, 2);
        setBindValue(stmt, 4, melt_pt, 3, 453.85);

        setBindValue(stmt, 5, nbr_valence, 1, 1);
        setBindNull(stmt, 5, 2);
        setBindValue(stmt, 5, nbr_valence, 3, 1);

        executeStmt(stmt, 3);

        // second batch

        setBindValue(stmt, 1, atom_nbr, 1, 4);
        setBindValue(stmt, 1, atom_nbr, 2, 5);

        setBindValue(stmt, 2, elem_name, 1, "Beryllium");
        setBindValue(stmt, 2, elem_name, 2, "Boron");

        setBindValue(stmt, 3, is_metal, 1, "yes");
        setBindNull(stmt, 3, 2);

        setBindValue(stmt, 4, melt_pt, 1, 1560.15);
        setBindValue(stmt, 4, melt_pt, 2, 2573.15);

        setBindValue(stmt, 5, nbr_valence, 1, 2);
        setBindValue(stmt, 5, nbr_valence, 2, 2);

        executeStmt(stmt, 2);

        // select inserted data

        stmt.Execute("select * from ptable");
        auto rs = stmt.GetResultset();
        while (rs++)
        {
            std::cout
                << std::left << std::setw(10) << rs.Get<std::string>(1)
                << std::left << std::setw(15) << rs.Get<std::string>(2)
                << std::left << std::setw(10) << rs.Get<std::string>(3)
                << std::left << std::setw(10) << rs.Get<std::string>(4)
                << std::left << std::setw(10) << rs.Get<std::string>(5)
                << std::endl;
        }

        con.Commit();

    }
    catch (Exception& ex)
    {
        ocout << ex.what() << oendl;
    }

    Environment::Cleanup();
}
```


```bash
row inserted : 3
row inserted : 2
1         Hydrogen                 14        1
2         Helium
3         Lithium        yes       454       1
4         Beryllium      yes       1560      2
5         Boron                    2573      2
```

**Comment by bangusi** [comment]
Thanks. So when doing repeated batch inserts you should always call `Statement::SetDataNull(false, elemIndex)`  unless you are certain your data contains no nulls.

One more question which is more efficient 
`Statement::GetBind(bind_index)`
or
`Statement::GetBind(bind_name)`
or doesn't matter?

**Comment by bangusi** [comment]
> arf, You're right, it is because, once you've called, BindInfo::SetDataNull(true, pos), this information is retained. You must revert it with BindInfo::SetDataNull(false, pos)

Was this by design?

**Comment by vrogier** [comment]
Always setting back to not null is the easy way, you can also do it only when you know that you need it.

Regarding the difference between the overloads of GetBind(), the one using index is, in the absolute, more efficient as the one using the name does a lookup into a hash map.
But depending your needs and the application, the difference might be not noticeable (unless you're doing heavy load computing)

About the need to revert back to not null once you've set null, it is about how Oracle binds works.
When you bind host data, you also bind "indicators" (variable saysing if the host data is null or not null) as there is not concept of NULL in C or C++.
Thus, Oracle lets you bind, for example, pass an array of int values for binding an array of int + an array of indicator providing the NULL information of each element in the array.
These indicators are used during the while execution time.
When calling BindInfo::SetDataNull(), OCILIB is just setting the value of the element indicator.
Oracle uses this value to find out if the value is null.
Thus, if you set it to NULL at some point, Oracle will think that element is null unless we update the value. 
Indicators are not reset to not null after each execution, first because binds can be out bind. In this case, it is the Oracle client that updates the value of the indicator to let you know that the out bind is NULL.
